pp_hot.c

```diff
@@ -3205,31 +3205,42 @@ PP(pp_match)
     /* push captures on stack */
 
     {
-        const I32 nparens = RXp_NPARENS(prog);
-        I32 i = (global && !nparens) ? 1 : 0;
+        const I32 logical_nparens = RXp_LOGICAL_NPARENS(prog);
+        I32 i = (global && !logical_nparens) ? 1 : 0;
+        I32 *l2p = RXp_LOGICAL_TO_PARNO(prog);
+        I32 *p2l_next = RXp_PARNO_TO_LOGICAL_NEXT(prog);
+        I32 paren, offs_start, offs_end;
 
         SPAGAIN;			/* EVAL blocks could move the stack. */
-        EXTEND(SP, nparens + i);
-        EXTEND_MORTAL(nparens + i);
-        for (i = !i; i <= nparens; i++) {
-            if (LIKELY(RXp_OFFS_VALID(prog,i)))
-            {
-                const I32 len = RXp_OFFS_END(prog,i) - RXp_OFFS_START(prog,i);
-                const char * const s = RXp_OFFS_START(prog,i) + truebase;
-                if ( UNLIKELY( len < 0 || len > strend - s)
-                )
-                    DIE(aTHX_ "panic: pp_match start/end pointers, i=%ld, "
-                        "start=%ld, end=%ld, s=%p, strend=%p, len=%" UVuf,
-                        (long) i, (long) RXp_OFFS_START(prog,i),
-                        (long)RXp_OFFS_END(prog,i), s, strend, (IV) len);
-                PUSHs(newSVpvn_flags(s, len,
-                    (DO_UTF8(TARG))
-                    ? SVf_UTF8|SVs_TEMP
-                    : SVs_TEMP)
-                );
-            } else {
-                PUSHs(sv_newmortal());
-            }
+        EXTEND(SP, logical_nparens + i);
+        EXTEND_MORTAL(logical_nparens + i);
+        for (i = !i; i <= logical_nparens; i++) {
+            paren = l2p ? l2p[i] : i;
+            do {
+                if (((offs_start = RXp_OFFS_START(prog, paren)) != -1) &&
+                    ((offs_end   = RXp_OFFS_END(prog, paren))   != -1))
+                {
+                    const I32 len = offs_end - offs_start;
+                    const char * const s = offs_start + truebase;
+                    if ( UNLIKELY( len < 0 || len > strend - s) ) {
+                        DIE(aTHX_ "panic: pp_match start/end pointers, paren=%" IVdf ", "
+                            "start=%" IVdf ", end=%" IVdf ", s=%p, strend=%p, len=%" IVdf,
+                            (IV)paren, (IV)offs_start, (IV)offs_end, s, strend, (IV)len);
+                        NOT_REACHED; /* NOTREACHED */
+                    }
+                    PUSHs(newSVpvn_flags(s, len,
+                        (DO_UTF8(TARG))
+                        ? SVf_UTF8|SVs_TEMP
+                        : SVs_TEMP)
+                    );
+                    break;
+                    NOT_REACHED; /* NOTREACHED */
```

Comment:
This and the similar below aren't useful.

The macro/comments are useful for telling the compiler/linter that this code can't be reached due to some sort of logic, but in both of these cases they're immediately preceded by an unconditional `break;` - the compiler or linter already knows the code can't be reached.